nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py

```diff
@@ -844,3 +844,383 @@ def list_available_models(cls) -> Optional[Dict[str, str]]:
+        """
+        result = []
+
+        return result
+
+    def sample_z(self, hidden, hidden_mask, return_ortho_loss=False):
+        """
+        Sample latent code z with reparameterization from bridge for
+        probabilistic latent variable models, or return value for
+        probabilistic models (seq2seq, seq2seq-br)
+        """
+        if self.model_type == "seq2seq":
+            # seq2seq
+            z = z_mean = hidden
+            z_logv = torch.zeros_like(hidden)
+            z_mask = hidden_mask
+            ortho_loss = 0.0
+        else:
+            # seq2seq-br, seq2seq-mim, seq2seq-vae
+
+            # project hidden to a fixed size bridge using k attention heads
+            res = self.att_bridge(
+                hidden=hidden,
+                hidden_mask=hidden_mask,
+                return_ortho_loss=return_ortho_loss,
+            )
+
+            if return_ortho_loss:
+                bridge_hidden, ortho_loss = res
+            else:
+                bridge_hidden = res
+
+            # all bottleneck models have mean
+            z_mean = self.hidden2latent_mean(bridge_hidden)
+
+            # sample posterior q(z|x) for MIM and VAE
+            if self.model_type in ["seq2seq-mim", "seq2seq-vae"]:
+                z_logv = self.hidden2latent_logv(bridge_hidden)
+                # avoid numerical instability for MIM
+                z_logv = z_logv.clamp_min(self.min_logv)
+                # sample z with reparameterization
+                e = torch.randn_like(z_mean)
+                z = e * torch.exp(0.5 * z_logv) + z_mean
+            else:
+                z_logv = torch.zeros_like(z_mean)
+                z = z_mean
+
+            # all steps in bottleneck bridge are used
+            z_mask = torch.ones(z.shape[0:2]).to(hidden_mask)
+
+        if return_ortho_loss:
+            return z, z_mean, z_logv, z_mask, ortho_loss
+        else:
+            return z, z_mean, z_logv, z_mask
+
+    @typecheck()
+    def forward(self, src, src_mask, tgt, tgt_mask, labels, train=True):
+        src_hiddens = self.encoder(
+            input_ids=src,
+            encoder_mask=src_mask,
+        )
+
+        # build posterior distribution q(x|z)
+        z, z_mean, z_logv, bridge_mask, ortho_loss = self.sample_z(
+            hidden=src_hiddens,
+            hidden_mask=src_mask,
+            return_ortho_loss=True,
+        )
+
+        # build decoding distribution
+        bridge_hiddens_dec = self.latent2hidden(z)
+
+        tgt_hiddens = self.decoder(
+            input_ids=tgt,
+            decoder_mask=tgt_mask,
+            encoder_embeddings=bridge_hiddens_dec,
+            encoder_mask=bridge_mask,
+        )
+
+        log_probs = self.log_softmax(hidden_states=tgt_hiddens)
+
+        recon_loss_fn = self.loss_fn if train else self.eval_loss_fn
+
+        if self.recon_per_token:
+            log_p_x_given_z_per_token = -recon_loss_fn(log_probs=log_probs, labels=labels)
```

Comment:
Since VAE and MIM losses are defined via log probabilities I wanted the code to closely reflect the mathematical formulation. I thought that might be useful for other researchers down the road.
Does that make sense?

Reply:
I will add all components to be logged in WANDB

Reply:
Done for training step. Still need to figure out how to log additional values to test/eval steps.

Reply:
Done for eval step. Will message you when code is tested.